dc/s2n-quic-dc/src/clock/macros.rs

```diff
@@ -0,0 +1,109 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+macro_rules! impl_clock {
+    () => {
+        use super::SleepHandle;
+        use core::{
+            fmt,
+            future::Future,
+            pin::Pin,
+            task::{Context, Poll},
+            time::Duration,
+        };
+        use pin_project_lite::pin_project;
+        use s2n_quic_core::{ready, time::Timestamp};
+        use tracing::trace;
+
+        #[derive(Clone, Debug)]
+        pub struct Clock(Instant);
+
+        impl Default for Clock {
+            #[inline]
+            fn default() -> Self {
+                Self(Instant::now())
+            }
+        }
+
+        impl s2n_quic_core::time::Clock for Clock {
+            #[inline]
+            fn get_time(&self) -> Timestamp {
+                let time = self.0.elapsed();
+                unsafe { Timestamp::from_duration(time) }
+            }
+        }
+
+        pin_project!(
+            pub struct Sleep {
+                clock: Clock,
+                #[pin]
+                sleep: time::Sleep,
+            }
+        );
+
+        impl s2n_quic_core::time::Clock for Sleep {
+            #[inline]
+            fn get_time(&self) -> Timestamp {
+                self.clock.get_time()
+            }
+        }
+
+        impl Future for Sleep {
+            type Output = ();
+
+            fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+                let this = self.project();
+                ready!(core::future::Future::poll(this.sleep, cx));
+                Poll::Ready(())
+            }
+        }
+
+        impl super::Sleep for Sleep {
+            #[inline]
+            fn update(self: Pin<&mut Self>, target: Timestamp) {
+                let target = unsafe { target.as_duration() };
+
+                // floor the delay to milliseconds to reduce timer churn
```

Comment:
Should we round up to 1ms minimum? I guess this is pre-existing...

Reply:
hmm good question... on the one hand, anything under a millisecond is going to be difficult to guarantee scheduling for. on the other, it might be surprising that a timer you just armed is already expired.

but, yes, it is pre-existing. i can add a TODO in my next PR for it.